sale_weekday_delivery/models/res_partner.py

```diff
@@ -0,0 +1,111 @@
+    delivery_schedule_tuesday = fields.Boolean(default=True, string="Tuesday")
+    delivery_schedule_wednesday = fields.Boolean(default=True, string="Wednesday")
+    delivery_schedule_thursday = fields.Boolean(default=True, string="Thursday")
+    delivery_schedule_friday = fields.Boolean(default=True, string="Friday")
+    delivery_schedule_saturday = fields.Boolean(default=True, string="Saturday")
+    delivery_schedule_sunday = fields.Boolean(default=True, string="Sunday")
+
+    @api.constrains(
+        "delivery_schedule",
+        "delivery_schedule_monday",
+        "delivery_schedule_tuesday",
+        "delivery_schedule_wednesday",
+        "delivery_schedule_thursday",
+        "delivery_schedule_friday",
+        "delivery_schedule_saturday",
+        "delivery_schedule_sunday",
+    )
+    def _check_delivery_schedule(self):
+        """Ensure at least one preferred day is defined"""
+        days_fields = ["delivery_schedule_%s" % d.lower() for d in calendar.day_name]
+        for partner in self:
+            if partner.delivery_schedule == "fix_weekdays" and not any(
+                [partner[d] for d in days_fields]
+            ):
+                raise ValidationError(
+                    _(
+                        "At least one preferred weekday must be defined if "
+                        "partner is set as 'Fixed week days' deliveries."
+                    )
+                )
+
+    def get_delivery_schedule_preferred_weekdays(self):
+        """List preferred weekdays"""
+        self.ensure_one()
+        if self.delivery_schedule == "direct":
+            return [d.lower() for d in calendar.day_name]
+        else:
+            return [
+                d.lower()
+                for d in calendar.day_name
+                if self["delivery_schedule_%s" % d.lower()]
+            ]
+
+    def _next_delivery_schedule_preferred_date(self, date_time=None):
+        """Returns next preferred date according to weekday preference"""
+        self.ensure_one()
+        if date_time is None:
+            date_time = fields.Datetime.today()
+        if not isinstance(date_time, fields.Datetime):
+            date_time = fields.Datetime.to_datetime(date_time)
+        preferred_weekdays = self.get_delivery_schedule_preferred_weekdays()
+        raw_date_weekday = date_time.weekday()
+        date_weekday = WEEKDAY_MAPPING.get(str(raw_date_weekday))
+        # Check if weekday is allowed
+        if (
+            date_weekday in preferred_weekdays
+            or self.delivery_schedule != 'fix_weekdays'
+        ):
+            return date_time
+        # Iterate over weekdays to find the next one being True
+        #  i = number of days to add until next valid weekday
+        #  wk_day = possible weekday number (to be converted by modulo 7 below)
+        for i, wk_day in enumerate(
+            range(raw_date_weekday + 1, raw_date_weekday + 7), 1
```

Comment:
you can use `tools.date_utils.date_range` :wink:

Reply:
Looking at it, I'm not sure it helps having a more readable code as we need must loop over weekdays instead of dates here... that's also why I took care of adding enough comments.
Anyway feel free to propose a PR if you find a more elegant way of doing this. :stuck_out_tongue_winking_eye: